_posts/2020-07-19-rate-limiting-with-resilience4j.md

```diff
@@ -0,0 +1,291 @@
+limiter.changeTimeoutDuration(Duration.ofSeconds(2));
+```
+
+This feature is useful if our rate limits vary based on time of day, for example - we could have a scheduled thread change these values. The new values won't affect the threads that are currently waiting for permissions.
+
+### Using `RateLimiter` and `Retry` Together
+
+Let's say we want to retry if we get a `RequestNotPermitted` exception since it is a transient error. We would create `RateLimiter` and `Retry` objects as usual. We then decorate a rate-limited `Supplier` and wrap it with a `Retry`:
+
+```java
+Supplier<List<Flight>> rateLimitedFlightsSupplier = RateLimiter.decorateSupplier(rateLimiter, 
+() -> service.searchFlights(request));
+Supplier<List<Flight>> retryingFlightsSupplier = Retry.decorateSupplier(retry, rateLimitedFlightsSupplier);
+```
+
+The sample output shows the request being retried for a `RequestNotPermitted` exception:
+
+```
+Searching for flights; current time = 17:10:09 218
+...
+[Flight{flightNumber='XY 765', flightDate='07/31/2020', from='NYC', to='LAX'}, ...]
+2020-07-27T17:10:09.484: Retry 'rateLimitedFlightSearch', waiting PT1S until attempt '1'. Last attempt failed with exception 'io.github.resilience4j.ratelimiter.RequestNotPermitted: RateLimiter 'flightSearchService' does not permit further calls'.
+Searching for flights; current time = 17:10:10 492
+...
+2020-07-27T17:10:10.494: Retry 'rateLimitedFlightSearch' recorded a successful retry attempt...
+[Flight{flightNumber='XY 765', flightDate='07/31/2020', from='NYC', to='LAX'}, ...]
+```
+
+The order in which we created the decorators is important. It would not work if we wrapped the `Retry` with the `RateLimiter`.
+
+## RateLimiter Events
+
+`RateLimiter` has an `EventPublisher` which generates events `RateLimiterOnSuccessEvent` and `RateLimiterOnFailureEvent` when calling a remote operation to indicate if acquiring permission was successful or not. We can listen for these events and log them, for example:
+
+```java
+RateLimiter limiter = registry.rateLimiter("flightSearchService");
+limiter.getEventPublisher().onSuccess(e -> System.out.println(e.toString()));
+limiter.getEventPublisher().onFailure(e -> System.out.println(e.toString()));
+```
+
+The sample output shows what's logged:
+
+```
+RateLimiterEvent{type=SUCCESSFUL_ACQUIRE, rateLimiterName='flightSearchService', creationTime=2020-07-21T19:14:33.127+05:30}
+... other lines omitted ...
+RateLimiterEvent{type=FAILED_ACQUIRE, rateLimiterName='flightSearchService', creationTime=2020-07-21T19:14:33.186+05:30}
+```
+
+## RateLimiter Metrics
+
+Suppose after implementing client-side throttling we find that the response times of our APIs have increased. This is possible because our request processing threads sometimes wait to get permission before calling the remote operation. How can we find out how often we are throttling our requests? How can we ensure that the values we've set on the `RateLimiterConfig` are working well?
```

Comment:
I don't understand the motivation to check how often we're throttling our requests when the response times have increased. Can you rephrase? I'm especially struggling with the second sentence

Reply:
Haha, it was a blind spot I guess. I had never noticed that. I had read somewhere that when you have long statements it's better to break the line after the operator so that a developer reading knowing that it continues, something like

int sum = a + 
                 b +
                 ...

I guess I had carried that over to "." as well. And you are right - I noticed after you mentioned that nearly everyone has the "." on the next line. I'll do the same next time. :)

Reply:
Agreed, couple of lines there were a bit confusing. I have made some changes to this and the next paragraph. Can you check if it reads better now?

Reply:
Yes, much clearer now! :)